crates/next-api/src/operation.rs

```diff
@@ -0,0 +1,137 @@
+    pub instrumentation: Option<InstrumentationOperation>,
+    pub pages_document_endpoint: VcOperation<Box<dyn Endpoint>>,
+    pub pages_app_endpoint: VcOperation<Box<dyn Endpoint>>,
+    pub pages_error_endpoint: VcOperation<Box<dyn Endpoint>>,
+}
+
+#[turbo_tasks::value_impl]
+impl EntrypointsOperation {
+    #[turbo_tasks::function]
+    pub async fn new(entrypoints: VcOperation<Entrypoints>) -> Result<Vc<Self>> {
+        let e = entrypoints.connect().await?;
+        Ok(Self {
+            routes: e
+                .routes
+                .iter()
+                .map(|(k, v)| (k.clone(), wrap_route(v, entrypoints)))
+                .collect(),
+            middleware: e.middleware.as_ref().map(|m| MiddlewareOperation {
+                endpoint: wrap(m.endpoint, entrypoints),
+            }),
+            instrumentation: e
+                .instrumentation
+                .as_ref()
+                .map(|i| InstrumentationOperation {
+                    node_js: wrap(i.node_js, entrypoints),
+                    edge: wrap(i.edge, entrypoints),
+                }),
+            pages_document_endpoint: wrap(e.pages_document_endpoint, entrypoints),
+            pages_app_endpoint: wrap(e.pages_app_endpoint, entrypoints),
+            pages_error_endpoint: wrap(e.pages_error_endpoint, entrypoints),
+        }
+        .cell())
+    }
+}
+
+fn wrap_route(route: &Route, entrypoints: VcOperation<Entrypoints>) -> RouteOperation {
+    match route {
+        Route::Page {
+            html_endpoint,
+            data_endpoint,
+        } => RouteOperation::Page {
+            html_endpoint: wrap(*html_endpoint, entrypoints),
+            data_endpoint: wrap(*data_endpoint, entrypoints),
+        },
+        Route::PageApi { endpoint } => RouteOperation::PageApi {
+            endpoint: wrap(*endpoint, entrypoints),
+        },
+        Route::AppPage(pages) => RouteOperation::AppPage(
+            pages
+                .iter()
+                .map(|p| AppPageRouteOperation {
+                    original_name: p.original_name.clone(),
+                    html_endpoint: wrap(p.html_endpoint, entrypoints),
+                    rsc_endpoint: wrap(p.rsc_endpoint, entrypoints),
+                })
+                .collect(),
+        ),
+        Route::AppRoute {
+            original_name,
+            endpoint,
+        } => RouteOperation::AppRoute {
+            original_name: original_name.clone(),
+            endpoint: wrap(*endpoint, entrypoints),
+        },
+        Route::Conflict => RouteOperation::Conflict,
+    }
+}
+
+#[turbo_tasks::function]
+fn wrap_endpoint(
+    endpoint: Vc<Box<dyn Endpoint>>,
+    op: VcOperation<Entrypoints>,
+) -> Vc<Box<dyn Endpoint>> {
+    let _ = op.connect();
+    endpoint
+}
+
+fn wrap(
+    endpoint: Vc<Box<dyn Endpoint>>,
+    op: VcOperation<Entrypoints>,
+) -> VcOperation<Box<dyn Endpoint>> {
+    VcOperation::new(wrap_endpoint(endpoint, op))
+}
```

Comment:
It's probably not worth it given the very limited number of `VcOperation`s in existence, but we could modify `#[turbo_tasks::function]` to allow returning a `VcOperation` directly.

Reply:
This is the current plan, but we're still figuring out exactly what this looks like in the macro.